src/main/java/org/killbill/billing/plugin/core/config/PaymentPluginTenantConfigurationHandler.java

```diff
@@ -0,0 +1,113 @@
+
+import javax.annotation.Nullable;
+
+import org.killbill.billing.osgi.libs.killbill.OSGIKillbillAPI;
+import org.killbill.billing.osgi.libs.killbill.OSGIKillbillLogService;
+import org.killbill.billing.plugin.api.notification.PluginConfigurationHandler;
+import org.killbill.billing.plugin.api.notification.PluginTenantConfigurable;
+import org.osgi.service.log.LogService;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.type.MapType;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+public abstract class PaymentPluginTenantConfigurationHandler<T> extends PluginConfigurationHandler {
+
+    private static ObjectMapper yamlObjectMapper = new ObjectMapper(new YAMLFactory());
+
+    private final Collection<UUID> configuredTenants = new HashSet<>();
+    private final PluginTenantConfigurable<T> pluginTenantConfigurable = new PluginTenantConfigurable<>();
+
+    private final ObjectReader yamlObjectReader;
+    private final String configurationKey;
+    private LogService osgiKillbillLogService;
+
+    public PaymentPluginTenantConfigurationHandler(final String pluginName,
+                                                   final OSGIKillbillAPI osgiKillbillAPI,
+                                                   final OSGIKillbillLogService osgiKillbillLogService,
+                                                   final String configurationKey) {
+        super(pluginName, osgiKillbillAPI, osgiKillbillLogService);
+        this.osgiKillbillLogService = osgiKillbillLogService;
+
+        this.configurationKey = configurationKey;
+        final MapType mapType = TypeFactory
+                .defaultInstance().constructMapType(Map.class, String.class, Map.class);
+        this.yamlObjectReader = yamlObjectMapper.readerFor(mapType);
+    }
+
+    public String getConfigurationKey() {
+        return configurationKey;
+    }
+
+    // Allow for later configuration
+    public void setDefaultConfigurable(final T defaultConfigurable) {
+        this.pluginTenantConfigurable.setDefaultConfigurable(defaultConfigurable);
+    }
+
+    @Override
+    protected void configure(@Nullable final UUID kbTenantId) {
+        final String rawConfiguration = getTenantConfigurationAsString(kbTenantId);
+        if (rawConfiguration != null) {
+            try {
+                final Map<String, Map<String, Object>> configObject = yamlObjectReader.readValue(rawConfiguration);
+                final T configurable = createConfigurable(configObject.getOrDefault(configurationKey, Collections
+                        .emptyMap()));
+                pluginTenantConfigurable.put(kbTenantId, configurable);
+            } catch (IOException e) {
+                osgiKillbillLogService.log(LogService.LOG_INFO, "Error while parsing YAML configuration, falling back to parsing Properties", e);
```

Comment:
I'm wondering if we should really have such a fallback. New plugins won't need to support both formats.

As mentioned in the original private PR, there are several cases where it can be ambiguous:

* `a: b\nc: d` (properties and YAML parsing would yield the same result)
* `a:\n - b\n - c` (valid properties parsed as `{"a"=>"", "-"=>"c"}` while YAML would yield `{"a"=>["b", "c"]}`)